packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -2108,19 +2110,29 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
     for (_RouteEntry entry in toBeDisposed) {
       entry.dispose();
     }
-
-    switch (SchedulerBinding.instance.schedulerPhase) {
-      case SchedulerPhase.persistentCallbacks:
-        SchedulerBinding.instance.addPostFrameCallback((Duration duration) {
-          overlay?.rearrange(_allRouteOverlayEntries);
-        });
-        break;
-      case SchedulerPhase.idle:
-      case SchedulerPhase.transientCallbacks:
-      case SchedulerPhase.midFrameMicrotasks:
-      case SchedulerPhase.postFrameCallbacks:
-        overlay?.rearrange(_allRouteOverlayEntries);
+    // Checks if any overlay entries are moved or added. Rearranges the overlay
+    // if so.
+    final List<OverlayEntry> newOverlayEntries = _allRouteOverlayEntries.toList();
```

Comment:
I don't fully understand this algorithm. If we wanted to base this on the overlayEntries, wouldn't we want to re-arrange whenever `listEquals(_allRouteOverlayEntries.toList(), _lastUpdatedOverlayEntries)` is false? Or am I missing something?

Also calling `_allRouteOverlayEntries.toList()` seems very expensive. Couldn't we run the algorithm on `_history` which basically determines the order returned by `_allRouteOverlayEntries`?

Another idea: When we call _flushHistory (this method) we should know whether we called _flushHistory because routes where added/removed/changed positions, which require re-arrangement of overlays. Could we give _flushHistory a `updateOverlays` param that's set to true by the caller of _flushHistory whenever we need to re-arrange the overlay?

Reply:
I am a little confused, do we want to skip only if there is no change in history other than lifecycle change?

Reply:
You're right to be confused. My comment above is incorrect. Sorry. 

Stepping back a little and looking at this more from a distance I noticed a thing:

I think the current design has it a little backwards: Instead of having the Route in charge of removing its overlays in its dispose, the Navigator should be in charge of that. That way, it is symmetric: The Navigator inserts the overlays in the Overlay and removes them again. The Route just provides them. 

My suggestion would be to do the following:

* Update all documentation on OverlayRoute (especially OverlayRoute.dispose) to indicate that the Navigator is in charge of removing the overlays.
* In the for loop above (line 2111) remove the overlays for each entry before disposing the entry.
* Remove this "do we need to rearrange the overlay" detection algorithm completely and just call `overlay.rearrange`
* Give _flushHistoryUpdates an optional `rearrangeOverlay` argument that defaults to true.
* In Navigator.removeRoute call _flushHistroyUpdates with rearrangeOverlays set to false since it never requires a rearrangement of overlays.

What do you think?

Again, sorry for the confusion and for having this backwards...